File: betamax-espresso/src/main/kotlin/software/betamax/android/PermissionRule.java
package software.betamax.android;

import android.support.test.rule.ActivityTestRule;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class PermissionRule implements TestRule {
  private final TapeDirectories tapeDirectories;
  private final ActivityTestRule<?> activityTestRule;

  public PermissionRule(TapeDirectories tapeDirectories, ActivityTestRule<?> activityTestRule) {
    this.tapeDirectories = tapeDirectories;
    this.activityTestRule = activityTestRule;
  }

  @Override public Statement apply(final Statement statement, Description description) {
    return new Statement() {
      @Override public void evaluate() throws Throwable {
        tapeDirectories.grantPermissionsIfNeeded(activityTestRule.getActivity());
        statement.evaluate();
      }
    };
  }
}

File: walkman-espresso/src/main/kotlin/walkman/PermissionRule.kt
package walkman

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class PermissionRule(
    private val configuration: WalkmanConfig) : TestRule {

  override fun apply(statement: Statement, description: Description): Statement {
    return object : Statement() {
      @Throws(Throwable::class) override fun evaluate() {
        val tapeRoot = configuration.tapeRoot
        if (tapeRoot is AndroidTapeRoot) {
          tapeRoot.grantPermissionsIfNeeded()
        } else {
          throw IllegalArgumentException("TapeRoot needs to be an instance of AndroidTapeRoot")
        }
        statement.evaluate()
      }
    }
  }
}
